src/components/icon/icon-registry.ts

```diff
@@ -0,0 +1,370 @@
+      return Observable.of(namedIcon);
+    }
+    // Not found in any cached icon sets. If there are icon sets with URLs that we haven't
+    // fetched, fetch them now and look for iconName in the results.
+    const iconSetFetchRequests: Observable<SVGElement>[] = iconSetConfigs
+        .filter(iconSetConfig => !iconSetConfig.svgElement)
+        .map(iconSetConfig =>
+            this._loadSvgIconSetFromConfig(iconSetConfig)
+                .catch((err: any, source: any, caught: any): Observable<SVGElement> => {
+                  // Swallow errors fetching individual URLs so the combined Observable won't
+                  // necessarily fail.
+                  console.log(`Loading icon set URL: ${iconSetConfig.url} failed: ${err}`);
+                  return Observable.of(null);
+                })
+                .do((svg: SVGElement) => {
+                  // Cache SVG element.
+                  if (svg) {
+                    iconSetConfig.svgElement = svg;
+                  }
+                }));
+    // Fetch all the icon set URLs. When the requests complete, every IconSet should have a
+    // cached SVG element (unless the request failed), and we can check again for the icon.
+    return Observable.forkJoin(iconSetFetchRequests)
+        .map((ignoredResults: any) => {
+          const foundIcon = this._extractIconWithNameFromAnySet(name, iconSetConfigs);
+          if (!foundIcon) {
+            throw new MdIconNameNotFoundException(name);
+          }
+          return foundIcon;
+        });
+  }
+
+  /**
+   * Searches the cached SVG elements for the given icon sets for a nested icon element whose "id"
+   * tag matches the specified name. If found, copies the nested element to a new SVG element and
+   * returns it. Returns null if no matching element is found.
+   */
+  private _extractIconWithNameFromAnySet(iconName: string, iconSetConfigs: SvgIconConfig[]):
+      SVGElement {
+    // Iterate backwards, so icon sets added later have precedence.
+    for (let i = iconSetConfigs.length - 1; i >= 0; i--) {
+      const config = iconSetConfigs[i];
+      if (config.svgElement) {
+        const foundIcon = this._extractSvgIconFromSet(config.svgElement, iconName, config);
+        if (foundIcon) {
+          return foundIcon;
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Loads the content of the icon URL specified in the SvgIconConfig and creates an SVG element
+   * from it.
+   */
+  private _loadSvgIconFromConfig(config: SvgIconConfig): Observable<SVGElement> {
+    return this._fetchUrl(config.url)
+        .map(svgText => this._createSvgElementForSingleIcon(svgText, config));
+  }
+
+  /**
+   * Loads the content of the icon set URL specified in the SvgIconConfig and creates an SVG element
+   * from it.
+   */
+  private _loadSvgIconSetFromConfig(config: SvgIconConfig): Observable<SVGElement> {
+    return this._fetchUrl(config.url)
+        .map((svgText) => this._svgElementFromString(svgText));
```

Comment:
Yeah, that felt sketchy to me also. Although I don't think it's any worse than ngMaterial 1.x.